backend/models/sequence.js

```diff
@@ -19,16 +19,102 @@
 const db = require("../handler/db");
 const responseCodes = require("../response_codes.js");
 const utils = require("../utils");
+const yup = require("yup");
 
+const { update: updateGroup } = require("./group");
 const History = require("./history");
 const FileRef = require("./fileRef");
 const { getRefNodes } = require("./ref");
 const { getDefaultLegendId } = require("./modelSetting");
+const View = new (require("./view"))();
+const { cleanViewpoint, createViewpoint } = require("./viewpoint");
 
-const activityCol = (modelId) => `${modelId}.activities`;
 const sequenceCol = (modelId) => `${modelId}.sequences`;
 const legendCol = (modelId) => `${modelId}.sequences.legends`;
 
+const findMinMaxDate =  (frames) => {
+	let min = frames[0].dateTime;
+	let max = frames[0].dateTime;
+
+	frames.forEach((frame) => {
+		if (frame.dateTime < min) {
+			min = frame.dateTime;
+		}
+		if (frame.dateTime > max) {
+			max = frame.dateTime;
+		}
+	});
+
+	return {min, max};
+};
+
+const byDateTime = (frameA, frameB) => frameA.dateTime - frameB.dateTime;
+
+const notEmpty = (obj) => Object.keys(obj).length > 0;
+
+// Viewpoint checks
+const viewpointSchema = yup.object().test((viewpoint) => {
+	if (!viewpoint) {
+		return true;
+	}
+
+	const {transformation_group_ids, transformation_groups } = viewpoint;
+
+	// The viewpoints for sequences CANT have transformations groups
+	if (transformation_group_ids || transformation_groups) {
+		return false;
+	}
+
+	return true;
+});
+
+const frameSchema = yup.object().shape({
+	dateTime: yup.number().required(),
+	viewpoint: viewpointSchema,
+	viewId: yup.string()
+}).noUnknown().test(({viewId, viewpoint}) => {
+	// The frame must have either viewpoint or viewId, not both
+	return (viewId && !viewpoint) || (!viewId && viewpoint);
+});
+
+const nameSchema = yup.string().min(1).max(30);
+
+const sequenceSchema = yup.object().shape({
+	name: nameSchema.required(),
+	startDate: yup.number(),
+	endDate: yup.number(),
+	rev_id: yup.string(),
+	frames: yup.array().of(frameSchema).min(1).required()
+}).noUnknown().test(({startDate, endDate, frames}) => {
+	if ((startDate !== undefined || endDate !== undefined) && frames) {
```

Comment:
Can frames be undefined? does yup still run this test if it failed the `min(1).required()`?

I'm thinking that because L94 onwards could otherwise be reduced to this:
```suggestion
	return !(startDate !== undefined && min > startDate ||endDate !== undefined && max < endDate)
```

Reply:
It could be undefined with sequenceEditSchema, It "inherited" from sequenceSchema. But I changed to  yup.object() so is no longer possible.

TBH I think  the change would be a little hard to understand at first glance. I would get rid only of the frames thing.

Reply:
The frames can be undefined when is checking from  sequenceEditSchema